File: AalsiTrader/ViewModels/ScreenerViewModel.swift
import Foundation

@Observable
final class ScreenerViewModel {
    var stocks: [SmartMoneyStock] = []
    var isLoading = false
    var error: String?

    var searchText = ""
    var signalFilter: StockSignal?
    var structureFilter: MarketStructure?
    var sortField: SortField = .confidence
    var sortAscending = false

    var chartData: [CandleData] = []
    var isLoadingChart = false

    // MARK: - Computed

    var filteredStocks: [SmartMoneyStock] {
        var result = stocks

        if !searchText.isEmpty {
            result = result.filter { $0.symbol.localizedCaseInsensitiveContains(searchText) }
        }
        if let signalFilter {
            result = result.filter { $0.signal == signalFilter }
        }
        if let structureFilter {
            result = result.filter { $0.structure == structureFilter }
        }

        return result.sorted { a, b in
            let comparison: Bool
            switch sortField {
            case .symbol: comparison = a.symbol < b.symbol
            case .price: comparison = a.price < b.price
            case .trendStrength: comparison = a.trendStrength < b.trendStrength
            case .confidence: comparison = a.confidence < b.confidence
            case .rsi: comparison = a.rsi < b.rsi
            case .momentum: comparison = (a.momentum5d ?? 0) < (b.momentum5d ?? 0)
            }
            return sortAscending ? comparison : !comparison
        }
    }

    var bullishCount: Int { stocks.filter { $0.signal == .BUY }.count }
    var bearishCount: Int { stocks.filter { $0.signal == .SELL }.count }
    var strongCount: Int { stocks.filter { $0.confidence >= 70 }.count }

    // MARK: - Actions

    func loadStocks(force: Bool = false) async {
        isLoading = true
        error = nil
        do {
            let response = try await ScreenerService.fetchStocks(force: force)
            stocks = response.stocks
            isLoading = false
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func loadChart(symbol: String) async {
        isLoadingChart = true
        do {
            chartData = try await ScreenerService.fetchChart(symbol: symbol)
            isLoadingChart = false
        } catch {
            isLoadingChart = false
        }
    }

    func clearFilters() {
        searchText = ""
        signalFilter = nil
        structureFilter = nil
    }
}

enum SortField: String, CaseIterable {
    case symbol = "Symbol"
    case price = "Price"
    case trendStrength = "Trend"
    case confidence = "Confidence"
    case rsi = "RSI"
    case momentum = "Momentum"
}

File: AalsiTraderAndroid/app/src/main/java/com/aalsitrader/android/viewmodel/ScreenerViewModel.kt
package com.aalsitrader.android.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.aalsitrader.android.model.*
import com.aalsitrader.android.repository.ScreenerRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class ScreenerViewModel : ViewModel() {
    private val repository = ScreenerRepository()

    private val _stocks = MutableStateFlow<List<SmartMoneyStock>>(emptyList())
    val stocks: StateFlow<List<SmartMoneyStock>> = _stocks.asStateFlow()

    private val _chartData = MutableStateFlow<List<CandleData>>(emptyList())
    val chartData: StateFlow<List<CandleData>> = _chartData.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    private val _signalFilter = MutableStateFlow<StockSignal?>(null)
    val signalFilter: StateFlow<StockSignal?> = _signalFilter.asStateFlow()

    private val _sortBy = MutableStateFlow(ScreenerSort.CONFIDENCE)
    val sortBy: StateFlow<ScreenerSort> = _sortBy.asStateFlow()

    init {
        loadStocks()
    }

    fun loadStocks(force: Boolean = false) {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            try {
                _stocks.value = repository.getStocks(force)
            } catch (e: Exception) {
                _error.value = e.message
            }
            _isLoading.value = false
        }
    }

    fun loadChart(symbol: String) {
        viewModelScope.launch {
            try {
                _chartData.value = repository.getChart(symbol)
            } catch (_: Exception) {
                _chartData.value = emptyList()
            }
        }
    }

    fun setSearchQuery(query: String) { _searchQuery.value = query }
    fun setSignalFilter(signal: StockSignal?) { _signalFilter.value = if (_signalFilter.value == signal) null else signal }
    fun setSortBy(sort: ScreenerSort) { _sortBy.value = sort }
    fun clearError() { _error.value = null }

    fun filteredStocks(): List<SmartMoneyStock> {
        var result = _stocks.value
        val query = _searchQuery.value
        if (query.isNotBlank()) {
            result = result.filter { it.symbol.contains(query, ignoreCase = true) }
        }
        _signalFilter.value?.let { filter ->
            result = result.filter { it.signal == filter }
        }
        result = when (_sortBy.value) {
            ScreenerSort.CONFIDENCE -> result.sortedByDescending { it.confidence }
            ScreenerSort.TREND_STRENGTH -> result.sortedByDescending { it.trendStrength }
            ScreenerSort.RSI -> result.sortedBy { it.rsi }
            ScreenerSort.PRICE -> result.sortedByDescending { it.price }
        }
        return result
    }
}

enum class ScreenerSort { CONFIDENCE, TREND_STRENGTH, RSI, PRICE }
